include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+
+  ~ArgumentType() = default;
+
+  ArgumentType(const ArgumentType &) = default;
+  ArgumentType &operator=(const ArgumentType &) = default;
+
+  ArgumentType(ArgumentType &&) = default;
+  ArgumentType &operator=(ArgumentType &&) = default;
+
+  bool verify() const debug_function { return Type.verify(); }
+
+  bool operator==(const ArgumentType &) const = default;
+};
+INTROSPECTION_NS(flatc, ArgumentType, Pos, Type);
+
+/// \brief KeyedObjectTraits for flatc::ArgumentType based on its position
+template<>
+struct KeyedObjectTraits<flatc::ArgumentType> {
+
+  static std::size_t key(const flatc::ArgumentType &Val) { return Val.Pos; }
+
+  static flatc::ArgumentType fromKey(const std::size_t &Pos) {
+    return flatc::ArgumentType(Pos);
+  }
+};
+
+/// \brief A function pointer type in flatc. Function pointers are actually
+/// typedefs.
+struct flatc::FunctionPtrType : public flatc::Type {
+  static constexpr const char *Tag = "!flatc::FunctionPtrType";
+
+  QualifiedType ReturnType;
+  SortedVector<ArgumentType> ArgumentTypes;
+  bool IsVariadic;
+  // FIXME: eventually we should do something about ABI;
+
+  FunctionPtrType(llvm::StringRef NameRef) :
+    Type(NameRef, TypeKind::FunctionPtr),
+    ReturnType(),
+    ArgumentTypes(),
+    IsVariadic(false) {}
+
+  FunctionPtrType() : FunctionPtrType("") {}
+  ~FunctionPtrType() = default;
+
+  FunctionPtrType(const FunctionPtrType &) = default;
+  FunctionPtrType &operator=(const FunctionPtrType &) = default;
+
+  FunctionPtrType(FunctionPtrType &&) = default;
+  FunctionPtrType &operator=(FunctionPtrType &&) = default;
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::FunctionPtr;
+  }
+
+private:
+  // Kind must be TypeKind::FunctionPtr; ReturnType and all the ArgumentTypes
+  // must verify; ArgumentTypes may contain the void type only if there is one
+  // argument type and the function is not variadic
+  bool verifyDerived() const debug_function;
+
+  friend struct flatc::Type;
+};
+INTROSPECTION_NS(flatc,
+                 FunctionPtrType,
+                 UUID,
+                 Name,
+                 Kind,
+                 ReturnType,
+                 ArgumentTypes,
+                 IsVariadic);
+
+// Make flatc::TypeKind yaml-serializable, required for making flatc::Type
+// yaml-serializable as well
+template<>
+struct llvm::yaml::ScalarEnumerationTraits<flatc::TypeKind::Values> {
+  static void enumeration(IO &io, flatc::TypeKind::Values &Val) {
+    io.enumCase(Val, "TypeKind::Invalid", flatc::TypeKind::Invalid);
+    io.enumCase(Val, "TypeKind::Primitive", flatc::TypeKind::Primitive);
+    io.enumCase(Val, "TypeKind::Enum", flatc::TypeKind::Enum);
+    io.enumCase(Val, "TypeKind::Typedef", flatc::TypeKind::Typedef);
+    io.enumCase(Val, "TypeKind::Struct", flatc::TypeKind::Struct);
+    io.enumCase(Val, "TypeKind::Union", flatc::TypeKind::Union);
+    io.enumCase(Val, "TypeKind::FunctionPtr", flatc::TypeKind::FunctionPtr);
```

Comment:
```suggestion
    io.enumCase(Val, "Invalid", flatc::TypeKind::Invalid);
    io.enumCase(Val, "Primitive", flatc::TypeKind::Primitive);
    io.enumCase(Val, "Enum", flatc::TypeKind::Enum);
    io.enumCase(Val, "Typedef", flatc::TypeKind::Typedef);
    io.enumCase(Val, "Struct", flatc::TypeKind::Struct);
    io.enumCase(Val, "Union", flatc::TypeKind::Union);
    io.enumCase(Val, "FunctionPtr", flatc::TypeKind::FunctionPtr);
```

Also, can you push this close to the definition of the `enum`?

Reply:
Will do.

Reply:
Done